### PR TITLE
Add Spinnaker support

### LINKED ISSUE #122: Support for FLIR Blackfly via Spinnaker SDK

### PR DESCRIPTION
**Pull request by g3force**
Resolves #122 

There is still an open issue that I reported to ptgrey - except from this, ready to be reviewed.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,10 @@
 project(ssl-vision)
 set(OpenGL_GL_PREFERENCE LEGACY)
-set(USE_NEURAL_CALIBRATION FALSE CACHE BOOL "Enable neural network calibration (requires OpenCV 2, not compatible to later versions)")
-set(USE_DC1394 TRUE CACHE BOOL "Compile with DC1394 (firewire cameras)")
 set(USE_QT5 FALSE CACHE BOOL "Use Qt5 instead of Qt4")
-set(USE_mvIMPACT FALSE CACHE BOOL "Compile with mvImpact (bluefox cameras)")
+set(USE_NEURAL_CALIBRATION FALSE CACHE BOOL "Enable neural network calibration (requires OpenCV 2, not compatible to later versions)")
```
Comment (joydeep-b) [inline CMakeLists.txt]: Should be "compatible *with*"

File: src/app/capture_thread.cpp [hunk 2]
```diff
@@ -87,6 +87,12 @@ CaptureThread::CaptureThread(int cam_id)
   captureBlueFox3 = new CaptureBlueFox3(bluefox3,camId);
 #endif
 
+#ifdef SPINNAKER
+  captureModule->addItem("Spinnaker");
+  settings->addChild( (VarType*) (spinnaker = new VarList("Spinnaker")));
```
Comment (joydeep-b) [inline src/app/capture_thread.cpp]: Recommend splitting up into two lines for readability.
```cpp
spinnaker = new VarList("Spinnaker");
settings->addChild(reinterpret_cast<VarType*>(spinnaker));
```

File: src/shared/capture/capture_spinnaker.cpp [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+#endif
+{
+  cam_id = (unsigned int) default_camera_id;
+  is_capturing = false;
+
+  #ifndef VDATA_NO_QT
+    mutex.lock();
+  #endif
+
+  settings->addChild(capture_settings = new VarList("Capture Settings"));
```
Comment (joydeep-b) [inline src/shared/capture/capture_spinnaker.cpp]: Please split into two statements for clarity.

File: src/shared/capture/capture_spinnaker.cpp [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+  settings->addChild(dcam_parameters  = new VarList("Camera Parameters"));
```
Comment (joydeep-b) [inline src/shared/capture/capture_spinnaker.cpp]: Ditto.

File: src/shared/capture/capture_spinnaker.cpp [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+  //=======================CAPTURE SETTINGS==========================
+  capture_settings->addChild(v_cam_bus          = new VarInt("cam idx",default_camera_id));
```
Comment (joydeep-b) [inline src/shared/capture/capture_spinnaker.cpp]: Ditto.

File: src/shared/capture/capture_spinnaker.cpp [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+  capture_settings->addChild(v_colorout         = new VarStringEnum("color mode", Colors::colorFormatToString(COLOR_RAW8)));
```
Comment (joydeep-b) [inline src/shared/capture/capture_spinnaker.cpp]: Ditto.

File: src/shared/capture/capture_spinnaker.cpp (hunk 3 above)
Comment (g3force) [reply src/shared/capture/capture_spinnaker.cpp]: I agree that it looks nicer when splitted. On the other hand, all capture classes do it like this. I wanted to be consistent with the rest of the vision code.

**Comment by g3force** [comment]
> Looks fine, some small changes requested.
> What is the Spinnaker bug that we are waiting on?

When using the StreamBufferHandlingMode "First Only", reading images from the camera blocks after a short time for ever. spinview (the Test Software of the SDK) does the same.
The "First Only" Mode would be the ideal one for us.
I'm currently in contact with the support of ptgrey.

File: src/shared/capture/capture_spinnaker.cpp (hunk 3 above)
Comment (joydeep-b) [reply src/shared/capture/capture_spinnaker.cpp]: Okay, let's go for consistency.

**Comment by joydeep-b** [comment]
> The "First Only" Mode would be the ideal one for us.
> I'm currently in contact with the support of ptgrey.

Is it possible to reduce the buffer size? Then we should be able to reduce the effect of the delay.

**Comment by g3force** [comment]
> > The "First Only" Mode would be the ideal one for us.
> > I'm currently in contact with the support of ptgrey.
> 
> Is it possible to reduce the buffer size? Then we should be able to reduce the effect of the delay.

Yes, the minimum buffer size is 3. But there was still a noticeable delay compared to the FirstOnly mode, as far as I could see.
But I'll test this again in more detail this evening.

**Comment by g3force** [comment]
I made the relevant streaming parameters changeable in the UI and made 'OldestFirst' the default, which is stable at the moment.
When ssl-vision runs at full fps (not dropping frames), there is no noticeable latency, regardless of the streaming mode. 
I artificially reduced the processing frame rate by putting a usleep(30000) right before pCam->getNextImage(). This drops the frame rate to <30fps and there is a significant latency for all modes except FirstOnly.

I'd say the PR is ready, but default parameters might be changed if the problem with the SDK gets solved.

**Comment by joydeep-b** [comment]
Okay, the usleep was just for the test, correct?

**Comment by g3force** [comment]
> Okay, the usleep was just for the test, correct?

Yes, to simulate overload.

**Pull request merged**